### PR TITLE
MSC3244 room capabilities implementation

### PR DESCRIPTION
**Pull request by t3chguy**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

Example response:

```json
{
    "capabilities": {
        "m.room_versions": {
            "default": "6",
            "available": {
                "1": "stable",
                "2": "stable",
                "3": "stable",
                "4": "stable",
                "5": "stable",
                "6": "stable",
                "7": "stable",
                "org.matrix.msc2176": "unstable",
                "org.matrix.msc3083": "unstable"
            },
            "org.matrix.msc3244.room_capabilities": {
                "knock": {
                    "preferred": "7",
                    "support": [
                        "7"
                    ]
                },
                "restricted": {
                    "preferred": null,
                    "support": [
                        "org.matrix.msc3083"
                    ]
                }
            }
        },
        "m.change_password": {
            "enabled": true
        }
    }
}

### CODE REVIEW

File: tests/rest/client/v2_alpha/test_capabilities.py [hunk 1]
```diff
@@ -102,3 +102,49 @@ def test_get_change_password_capabilities_password_disabled(self):
...
+        capabilities = channel.json_body["capabilities"]
+
+        self.assertEqual(channel.code, 200)
+        for capability, details in capabilities["m.room_versions"][
+            "org.matrix.msc3244.room_capabilities"
+        ].items():
+            if details["preferred"] is not None:
+                self.assertTrue(
+                    details["preferred"] in KNOWN_ROOM_VERSIONS,
+                    "" + details["preferred"],
```
Comment (clokep) [inline tests/rest/client/v2_alpha/test_capabilities.py:145]:
```suggestion
                    str(details["preferred"]),
```

File: tests/rest/client/v2_alpha/test_capabilities.py [hunk 2]
```diff
@@ -102,3 +102,49 @@ def test_get_change_password_capabilities_password_disabled(self):
...
+                )
+
+            self.assertTrue(len(capability["support"]) > 0)
+            for room_version in capability["support"]:
+                self.assertTrue(room_version in KNOWN_ROOM_VERSIONS, "" + room_version)
```
Comment (clokep) [inline tests/rest/client/v2_alpha/test_capabilities.py:150]: I think these both want to look at the details dictionary, not the capability name.

```suggestion
            self.assertGreater(len(details["support"]), 0)
            for room_version in details["support"]:
                self.assertTrue(room_version in KNOWN_ROOM_VERSIONS, str(room_version))
```

Although if we don't care about the capability name we could probably iterate over `capabilities["m.room_versions"]["org.matrix.msc3244.room_capabilities"].values()` instead.

File: tests/rest/client/v2_alpha/test_capabilities.py [hunk 3]
```diff
@@ -102,3 +102,49 @@ def test_get_change_password_capabilities_password_disabled(self):
...
+            )
+        )
+
+        channel = self.make_request("GET", self.url, access_token=access_token)
+        capabilities = channel.json_body["capabilities"]
+
+        self.assertEqual(channel.code, 200)
+        self.assertIsNone(
+            capabilities["m.room_versions"]["org.matrix.msc3244.room_capabilities"]
+        )
```
Comment (clokep) [inline tests/rest/client/v2_alpha/test_capabilities.py:122]:
```suggestion
        self.assertNotIn(
            "org.matrix.msc3244.room_capabilities", capabilities["m.room_versions"]
        )
```

File: synapse/api/room_versions.py [hunk 4]
```diff
@@ -208,5 +208,39 @@ class RoomVersions:
...
 }  # type: Dict[str, RoomVersion]
+
+
+@attr.s(slots=True, frozen=True)
+class RoomVersionCapability:
+    """An object which describes the unique attributes of a room version."""
+
+    identifier = attr.ib(type=str)  # the identifier for this capability
+    preferred_version = attr.ib(type=Optional[RoomVersion])
+    support_check_lambda = attr.ib(type=Callable[[RoomVersion], bool])
```
Comment (clokep) [inline synapse/api/room_versions.py:220]: We're starting to use native type hints in new code:

```suggestion
@attr.s(slots=True, frozen=True, auto_attribs=True)
class RoomVersionCapability:
    """An object which describes the unique attributes of a room version."""

    identifier: str  # the identifier for this capability
    preferred_version: Optional[RoomVersion]
    support_check_lambda: Callable[[RoomVersion], bool])
```

File: tests/rest/client/v2_alpha/test_capabilities.py (hunk 2 above)
Comment (t3chguy) [reply tests/rest/client/v2_alpha/test_capabilities.py:150]: 🤦 oh, yes

**Comment by clokep** [comment]
Note that complement failing is unrelated to this PR, see #10275.

**Comment by richvdh** [comment]
for links: this implements [MSC3244](https://github.com/matrix-org/matrix-doc/issues/3244).

**Review by richvdh — APPROVED** [review]
lgtm

**Pull request merged**